DoubleSpinWidget, SpinWidget: add units, make usage more consistent

zwim opened the pull request:
Make default button, returning the default values -> 1 code additions, 2 code deletions.

Update OK button (gray out) when default button is tapped to the default values (and not the then obsolete old value).

No updating the button is annoying, if you tab the default button and dismiss the dialog by tapping outside it, then the new values are applied, but the ok button is still black (and not as expected gray).
Or if you tap the default button, but you want to go back to your old value, because the button is disabled then.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9046)
<!-- Reviewable:end -->

hius07 commented:
What do you think about
```lua
local function onApplyCallback(left_value, right_value)
    self.left_value = left_value
    self.right_value = right_value
    self.callback(left_value, right_value)
    if self.keep_shown_on_apply then
        self:update()
    else
        self:onClose()
    end
end
```
for both 'Apply' and 'Apply default values' buttons?

zwim commented:
Yes, this would make the code a bit shorter and clearer.

hius07 commented:
This can be removed I think
https://github.com/koreader/koreader/blob/7c4bf29e5a8efe784f3e3f98e6d98b386c4f5036/frontend/apps/reader/modules/readerdevicestatus.lua#L198
https://github.com/koreader/koreader/blob/7c4bf29e5a8efe784f3e3f98e6d98b386c4f5036/plugins/statistics.koplugin/main.lua#L968

hius07 commented:
Minor changes to the comment needed
https://github.com/koreader/koreader/blob/9650acdde61414a417dc2e2d8dfd55ae0146107e/frontend/ui/widget/doublespinwidget.lua#L51

zwim commented:
And this is also ready frim my side.

hius07 commented:
I like it, thanks.
Just a minor question what is the best separator between default values?
https://github.com/koreader/koreader/blob/e35a103e8c1777599317a34673498562ea4fb4e1/frontend/ui/widget/doublespinwidget.lua#L187
https://github.com/koreader/koreader/blob/e35a103e8c1777599317a34673498562ea4fb4e1/frontend/apps/reader/modules/readertypography.lua#L379

zwim commented:
I tried not to change the separator. In the hyphenation limits there was an emdash, so I added the info with the emdash to the menu.
Default would be "/".
But in readerdevicestatus there is an emdash in the menu and a "/" in the default button, 
The statistics plugin uses "/" in the menu and the default button.

So maybe we could change the default text to an emdash?
What do you think?

Frenzie commented:
Some comparative screenshots would help. :-)

(But yeah, that's probably fine.)

zwim commented:
Currently, it looks like this. Should we change the separators to an emdash or a "/"?

Statistics:
![grafik](https://user-images.githubusercontent.com/36999612/166902246-96a2617b-ac1b-453b-9bd6-98cd1fec6d32.png)
![grafik](https://user-images.githubusercontent.com/36999612/166902331-2a36f086-b94c-486e-b0e0-3dd116323bb9.png)


Device-statistics:
![grafik](https://user-images.githubusercontent.com/36999612/166902461-bb546cb3-e005-488a-8787-054fcb1d9db5.png)
![grafik](https://user-images.githubusercontent.com/36999612/166902522-443bf7b3-f584-4b68-8923-9c8303990f8d.png)

Reader-typography:
![grafik](https://user-images.githubusercontent.com/36999612/166902664-92c33588-0280-4fb2-bbad-3bd59ef984e3.png)
![grafik](https://user-images.githubusercontent.com/36999612/166902745-b2501dfd-b8fc-466a-acff-1bc70897b634.png)

zwim commented:
I would say in the former two an emdash would be right, as it is showing a range.

In readertypography a "/" would be more adequate.

hius07 commented:
> The second one is not a range (but the inverse it), so also here a "/" would show that.

Yes, emdash is confusing here (and in the main menu item), it is **not** the range, mea culpa.

poire-z commented:
No objection to any of the proposed changes.
( `/` may also look like a ratio, `3 / 5`, so keep that in mind in some contexts).

I'm not really fresh on the spin widgets - and don't remember much about the `nil, nil` trick - I just remember in some contexts, mostly the CRE tunables, I wanted to avoid a auto-apply, when resetting the spins to the defaults, as it could generate a re-rendering and would have the user wait - when he only wanted to get the defaults to start over from nominal values and have to hit the -/+ keys less.

zwim commented:
> I wanted to avoid a auto-apply, when resetting the spins to the defaults, as it could generate a re-rendering and would have the user wait - when he only wanted to get the defaults to start over from nominal values and have to hit the -/+ keys less.

@poire-z I will look into this tonight.

hius07 commented:
> And the doublespinwidget would look like that:

Should there be a space between the value and the percent sign? Not a simple question...
https://en.wikipedia.org/wiki/Percent_sign
(Since there are different standards, let it be, in consistency with other units)

zwim commented:
@hius07 Ok to the percent sign, but we really have one exception it's the degree symbol `°`.

hius07 commented:
> when he only wanted to get the defaults to start over from nominal values and have to hit the -/+ keys less

That is the behaviour of the Default button in the SpinWidget.
I vaguely remember we have discussed why DSW applies it immediately.

poire-z commented:
We have discussed it around https://github.com/koreader/koreader/pull/8490#discussion_r758761489 (I quickly re-read it, not sure I understand what we ended up with :).

For hyphenation limits, on a week old koreader nightly, <kbd>Use language defaults</kbd> indeed applies immediately - which would not be my prefered behaviour, but ok. But there is a minor inconsistency: it applies immediately, but the <kbd>Apply</kbd> button then becomes black/enabled (it should stay gray/dimmed if the displayed values are the ones that are applied).
But I guess the most natural would be to not apply immediately, and get the <kbd>Apply</kbd> button black/enabled, so the user knows it has to hit it to have the displayed values applied.
(And may be that's what's this PR fixes, dunno, haven't really followed :)

zwim commented:
> For hyphenation limits, on a week old koreader nightly, Use language defaults indeed applies immediately - which would not be my prefered behaviour, but ok. But there is a minor inconsistency: it applies immediately, but the Apply button then becomes black/enabled (it should stay gray/dimmed if the displayed values are the ones that are applied).

I did not change the behavior of anything in this PR. I have just made the OK-Apply button grey-out consistend with what's happening in the code.
And at the end I have changed some menu entries and added the unit thing to the DSW, SW and numberpicker.

> But I guess the most natural would be to not apply immediately, and get the Apply button black/enabled, so the user knows it has to hit it to have the displayed values applied.

Right, but as said, the objective was not to change the code behavior, but only make visible what is really happening.

> (And may be that's what's this PR fixes, dunno, haven't really followed :)

I am afraid no. But with this PR we might realize similar quirks earlier (during development) in future.

hius07 commented:
> applies immediately - which would not be my prefered behaviour

If there are no reasons for the difference between SW and DSW default buttons, I would prefer applying on Apply button only (SW).

zwim commented:
As you wish :)
IIRC the reason is, that I haven't wanted to change the existing code behavior. 
I'll do it tomorrow.

poire-z commented:
Go ahead fixing/making everything consistent and behaving logically - the current state may not be the best, and as you now have a good overview of everything that use them, you're the best to do it.
Then, we'll test it and see if some cases feels odd, and we'll see if some special cases handling are needed.

I agree that if we now have a Apply button that visually says that what we see is not what is applied, we don't need to have "Apply defaults". A "Use defaults" (or other wording) meaning update the spin values with default values, and make the Apply button enabled if the additional action is required from the user to have them really applied.

hius07 commented:
> A "Use defaults" (or other wording)

Just "Default values:"
https://github.com/koreader/koreader/blob/9b9cfe29a49423f8ee850cd8b3a6d05edba1a270/frontend/ui/widget/spinwidget.lua#L126

zwim commented:
Wording it "Default value (s )" without a verb. IIRC we had that duscussions some time before.

But over the time I become more convinced this is right. I feels good.
A button without a verb, meaning just a noun, show that nothing is actually done. It is an info.
As a sideffect you may tap on it, and some spinners are turned.


"Apply " or "close " do something.

zwim commented:
DSW and SW use now a consistent "Default value(s) ..." button, which just select the default value(s), but does not apply them.

zwim commented:
One difference consists in the callback. DSW calls the callback with (self.left_value, self.right_value), so the values can be used directly in the callback. SW calls  the callback with (self), so the values can only be accessed with self.value.

I think the DSW is the more programmer friendly version (less keystrokes, more readable as you can call the variables by their right name in the callback). But changing all SW in the whole code to this behavior is really unfun work (there are 96 SpinWidgets in frontend, not counted the plugins). 
So I would prefer not change this.

poire-z commented:
> So I would prefer not change this.

Fine by me.

hius07 commented on frontend/apps/reader/modules/readertypography.lua:
```diff
@@ -367,23 +367,27 @@ When the book's language tag is not among our presets, no specific features will
                 -- the hyphenation changes happening
                 width_factor = 0.6,
                 default_values = true,
-                default_text = _("Use language defaults"),
```
Isn't it necessary to emphasize "**language** defaults"?

zwim commented:
Rebased and added some more units to SW. 

From my side this is ready now, nothing more to add.

Now it's time to search for quirks.

hius07 commented on frontend/ui/data/onetime_migration.lua:
```diff
@@ -387,6 +387,18 @@ if last_migration_date < 20220426 then
+            G_reader_settings:delSetting(old)
+        end
+    end
+    migrateSettingsName("highlight_long_hold_threshold_s", "highlight_long_hold_threshold_s")
```
Old and new the same.

hius07 commented on frontend/ui/widget/doublespinwidget.lua:
```diff
@@ -167,17 +198,26 @@ function DoubleSpinWidget:update(numberpicker_left_value, numberpicker_right_val
+            elseif self.unit ~= "" then
+                unit = "\xE2\x80\xAF" .. self.unit -- use Narrow No-Break Space (NNBSP) here
+            end
+        end
```
Is it worth to add to utils `getUnitString`? Have 3 times of this code.

zwim replied:
I would not add this to `util`, as it is really very special.

hius07 commented on frontend/apps/reader/modules/readerhighlight.lua:
```diff
@@ -397,7 +397,8 @@ function ReaderHighlight:addToMainMenu(menu_items)
             {
-                text = _("Highlight long-press interval"),
+                text = T(_("Highlight long-press interval: %1 s"),
+                    G_reader_settings:readSetting("highlight_long_hold_threshold_s", 3)),
```
Will the menu item be updated on changing the interval?

zwim replied:
No. I will change that.

poire-z commented on frontend/ui/widget/doublespinwidget.lua:
```diff
@@ -117,6 +123,13 @@ function DoubleSpinWidget:update(numberpicker_left_value, numberpicker_right_val
     right_widget.picker_updated_callback = function(value)
         self:update(left_widget:getValue(), value)
     end
+    local separator_widget = TextBoxWidget:new{
+        text = self.is_range and "–" or "",
+        alignment = "center",
+        face = self.title_face,
+        bold = true,
+        width = math.floor(math.min(self.screen_width, self.screen_height) * 0.02),
```
Feels a bit odd to have to use a TextBoxWidget, when everything else use a simpler TextWidget - but I guess you used that for the fixed width and centering. This fixed width computation feels also odd: we would probably be better with a TextWidget + some padding.
It's some UI finicky/tricky work you may not (yet) be confortable with, so I'll let this one pass :)

zwim commented:
It feels odd to show a value with leading "0" when a unit is present. `03 min` vs `3 min`.
So I will change the default precision to `%1d` if a unit is present. Tomorrow

hius07 commented on frontend/ui/screensaver.lua:
```diff
@@ -391,7 +391,9 @@ function Screensaver:setStretchLimit(touchmenu_instance)
-        default_value = 8, -- percent
+        default_value = 8,
+        unit = "%",
+        precission = "%1d",
```
precision

hius07 commented on frontend/apps/reader/modules/readerhighlight.lua:
```diff
@@ -408,15 +411,18 @@ If a touch is not released in this interval, it is considered a long-press. On d
                         value_step = 1,
                         value_hold_step = 5,
+                        unit = _("s"),
+                        precision = "%1d",
```
Not needed anymore.

Frenzie commented on plugins/autowarmth.koplugin/main.lua:
```diff
@@ -533,6 +533,7 @@ function AutoWarmth:getLocationMenu()
                 value_step = 10,
                 value_hold_step = 100,
+                -- @translators This is the unit for meter.
                 unit = _("m"),
```
Context to disambiguate is likely still required regardless while also being self-explanatory.

Frenzie commented on plugins/autowarmth.koplugin/main.lua:
```diff
@@ -532,6 +533,8 @@ function AutoWarmth:getLocationMenu()
                 wrap = false,
                 value_step = 10,
                 value_hold_step = 100,
+                -- @translators This is the unit of length in meters.
+                unit = _("m"),
```
Ah no, I meant more like this, sorry.
```suggestion
                unit = _C("Length", "m"),
```

zwim replied:
I have learn something new. Thanks

Frenzie replied:
Without a context all instances of a particular string are regarded as the same. With a context you can differentiate between different types of the same string. Some projects add a context to all strings.

In the translation interface the context can function similarly to a comment.

Frenzie commented on frontend/apps/reader/modules/readerdevicestatus.lua:
```diff
@@ -148,6 +148,7 @@ function ReaderDeviceStatus:addToMainMenu(menu_items)
                         value_min = 1,
                         value_max = 60,
                         default_value = 10,
+                        unit = _("min"),
```
Implicitly I also meant this one btw (extrema vs time).

Frenzie commented on frontend/apps/reader/modules/readerdevicestatus.lua:
```diff
@@ -240,6 +237,7 @@ High level threshold is checked when the device is charging.]]),
                         value_min = 1,
                         value_max = 60,
                         default_value = 5,
+                        unit = _("min"),
```
And that one.

Frenzie commented on frontend/apps/reader/modules/readerscrolling.lua:
```diff
@@ -162,6 +162,7 @@ Default value: %1 ms]]), scroll_activation_delay_default_ms),
                         value_max = 2000,
                         value_step = 100,
                         value_hold_step = 500,
+                        unit = _("ms"),
```
For ms and s a context isn't necessary exactly, but it can be useful to filter by context and in this instance as built-in documentation.

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -484,6 +486,8 @@ The interval value is in milliseconds and can range from 0 (0 seconds) to 2000 (
                         value_max = 2000,
                         value_step = 50,
                         value_hold_step = 200,
+                        -- @translators This is the time unit for milliseconds.
+                        unit = _("ms"),
```
Yeah, context instead of comment would work better in the translation UI imo.

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -535,6 +541,8 @@ The interval value is in milliseconds and can range from 100 (0.1 seconds) to 20
                         value_max = 2000,
                         value_step = 100,
                         value_hold_step = 500,
+                        -- @translators This is the time unit for milliseconds.
+                        unit = _("ms"),
```
.

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -561,6 +569,8 @@ The duration value is in milliseconds and can range from 100 (0.1 seconds) to 20
                         value_max = 2000,
                         value_step = 100,
                         value_hold_step = 500,
+                        -- @translators This is the time unit for milliseconds.
+                        unit = _("ms"),
```
.

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -587,6 +597,8 @@ The interval value is in milliseconds and can range from 100 (0.1 seconds) to 20
                         value_max = 2000,
                         value_step = 100,
                         value_hold_step = 500,
+                        -- @translators This is the time unit for milliseconds.
+                        unit = _("ms"),
```
.

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -612,6 +624,8 @@ The interval value is in milliseconds and can range from 100 (0.1 seconds) to 20
                         value_min = 100,
                         value_max = 2000,
                         value_step = 100,
+                        -- @translators This is the time unit for milliseconds.
+                        unit = _("ms"),
```
.

zwim commented:
Ready from my side.

zwim commented:
Ready to merge.

Frenzie commented on plugins/autowarmth.koplugin/main.lua:
```diff
@@ -532,6 +534,8 @@ function AutoWarmth:getLocationMenu()
                 wrap = false,
                 value_step = 10,
                 value_hold_step = 100,
+                -- @translators This is the unit of length in meters.
```
```suggestion
```

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -457,6 +458,8 @@ Higher values mean faster screen updates, but also use more CPU.]]),
                         value_max = 60.0,
                         value_step = 1,
                         value_hold_step = 15,
+                         -- @translators This is the frequency unit Hertz.
```
```suggestion
```

Frenzie commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -587,6 +598,8 @@ The interval value is in milliseconds and can range from 100 (0.1 seconds) to 20
                         value_max = 2000,
                         value_step = 100,
                         value_hold_step = 500,
+                        -- @translators This is the time unit for milliseconds.
```
```suggestion
```

zwim commented on frontend/util.lua:
```diff
@@ -203,18 +204,18 @@ function util.secondsToHClock(seconds, withoutSeconds, hmsFormat, withDays)
         end
         if hmsFormat then
             -- @translators This is the 'h' for hour, like in 1h30m30s. This is a duration.
```
Thanks

NiLuJe reviewed (approved):
> Reviewed 2 of 15 files at r1, 1 of 2 files at r5, 2 of 8 files at r7, 11 of 12 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9046)* status: all files reviewed, 28 unresolved discussions (waiting on @Frenzie, @hius07, and @zwim)



<!-- Sent from Reviewable.io -->

NiLuJe reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9046)* status: all files reviewed, 28 unresolved discussions (waiting on @Frenzie, @hius07, and @zwim)



<!-- Sent from Reviewable.io -->

poire-z reviewed (approved):
Again not a thorough review, but nothing against it - we'll see in the wild.

zwim commented on frontend/ui/data/onetime_migration.lua:
```diff
@@ -7,7 +7,7 @@ local lfs = require("libs/libkoreader-lfs")
 
 -- Date at which the last migration snippet was added
-local CURRENT_MIGRATION_DATE = 20220426
+local CURRENT_MIGRATION_DATE = 20220516
```
Yep, I have that in my agenda. But thanks anyway.

zwim commented:
@poire-z  I am pretty sure that the last commit (#[2fffea5](https://github.com/koreader/koreader/pull/9046/commits/2fffea54dd0933bd319aefc4f1c21b86109b13f9)) makes sense.It's edge case, but who knows. Can you confirm that before merging? Thanks

poire-z commented:
Yes, it makes sense to :free() the textwidgets if they are removed.
But it would look neater if you just added things when you need to have them :)
I think VerticalGroup (inheriting from WidgetContainer) allows for adding stuff to it after :new() until the first :getSize() is called, so this may work:

```lua
   local right_vertical_group = VerticalGroup:new{
        align = "center",
        right_widget,
    }
    if self.left_text ~= "" or self.right_text == "" then
        table.insert(right_vertical_group, 1, TextWidget:new{
            text = self.right_text,
            face = self.title_face,
            max_width = text_max_width,
        })
```

zwim commented:
To give something back (@poire-z, honor to you, and you are way better than the drill sergeant):

```
He's fast, He's lean
he's frightening 
```

Well neat rhymes almost to lean 
https://www.youtube.com/watch?v=R7SHhS7D4IA

P.S. @Frenzie I'll get you for sure. The drill seargeant will be never ever forgotten :)

poire-z commented:
It it works for you, it works for me :) feel free merging if you don't want to update the date again.
